gtsam/basis/Basis.h

```diff
@@ -0,0 +1,373 @@
+      return apply(F, H);  // might call apply in derived
+    }
+  };
+
+  /// Base class for functors below that calculates weights
+  class DerivativeFunctorBase {
+   protected:
+    typename DERIVED::Weights weights_;
+
+   public:
+    DerivativeFunctorBase(size_t N, double x)
+        : weights_(DERIVED::DerivativeWeights(N, x)) {}
+
+    DerivativeFunctorBase(size_t N, double x, double a, double b)
+        : weights_(DERIVED::DerivativeWeights(N, x, a, b)) {}
+
+    void print(const std::string& s = "") const {
+      std::cout << s << (s != "" ? " " : "") << weights_ << std::endl;
+    }
+  };
+
+  /// Given values f at the Chebyshev points, predict derivative at x
+  class DerivativeFunctor : protected DerivativeFunctorBase {
+   public:
+    // Used by FunctorizedFactor
+    using argument_type = typename DERIVED::Parameters;
+    using return_type = double;
+
+    DerivativeFunctor(size_t N, double x) : DerivativeFunctorBase(N, x) {}
+
+    DerivativeFunctor(size_t N, double x, double a, double b)
+        : DerivativeFunctorBase(N, x, a, b) {}
+
+    double apply(const typename DERIVED::Parameters& f,
+                 OptionalJacobian</*1xN*/ -1, -1> H = boost::none) const {
+      if (H) *H = this->weights_;
+      return (this->weights_ * f)(0);
+    }
+    /// c++ sugar
+    double operator()(const typename DERIVED::Parameters& f,
+                      OptionalJacobian</*1xN*/ -1, -1> H = boost::none) const {
+      return apply(f, H);  // might call apply in derived
+    }
+  };
+
+  /// Vector interpolation, e.g. given 3*N matrix yields 3-vector
+  template <int M>
+  class VectorDerivativeFunctor : protected DerivativeFunctorBase {
+   protected:
+    typedef Eigen::Matrix<double, M, 1> VectorM;
+    typedef Eigen::Matrix<double, M, -1> MatrixMN;
+    typedef Eigen::Matrix<double, /*MxMN*/ M, -1> Jacobian;
+    Jacobian H_;
+    void calculateJacobian() {
+      typedef Eigen::Matrix<double, M, M> MatrixM;
+      H_ = Eigen::kroneckerProduct(this->weights_, MatrixM::Identity());
+    }
+
+   public:
+    // Used by FunctorizedFactor
+    using argument_type = Matrix;
+    using return_type = Vector;
+
+    EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+    /// Default Constructor
+    VectorDerivativeFunctor(size_t N, double x) : DerivativeFunctorBase(N, x) {
+      calculateJacobian();
+    }
+
+    /// Constructor, with optional interval [a,b]
+    VectorDerivativeFunctor(size_t N, double x, double a, double b)
+        : DerivativeFunctorBase(N, x, a, b) {
+      calculateJacobian();
+    }
+
+    VectorM apply(const MatrixMN& F,
+                  OptionalJacobian</*MxMN*/ -1, -1> H = boost::none) const {
+      if (H) *H = H_;
+      return F * this->weights_.transpose();
+    }
+    /// c++ sugar
+    VectorM operator()(const MatrixMN& F, OptionalJacobian</*MxMN*/ -1, -1> H =
+                                              boost::none) const {
+      return apply(F, H);
```

Comment:
Formatting?

Reply:
Yeah this is old code, definitely needs some more work.